src/basecamp.cpp

```diff
@@ -23,24 +68,98 @@ std::string basecamp::board_name() const
     return string_format( _( "%s Board" ), name.c_str() );
 }
 
-std::string basecamp::save_data() const
+#include <stdio.h>
+void basecamp::define_camp( npc &p )
+{
+    const auto parse_expansion = []( const std::string &terrain ) {
+        expansion_data e;
+        int last_bar = terrain.find_last_of( '_' );
+        e.type = terrain.substr( 13, last_bar - 13 );
+        e.cur_level = std::stoi( terrain.substr( last_bar + 1 ) );
+        return e;
+    };
+    sort_points = p.companion_mission_points;
+    // purging the regions guarantees all entries will start with faction_base_
+    for( std::pair<std::string, tripoint> expansion : talk_function::om_building_region( p, 1,
+            true ) ) {
+        const std::string dir = talk_function::om_simple_dir( pos, expansion.second );
+        expansions[ dir ] = parse_expansion( expansion.first );
+        expansion_data &e = expansions[ dir ];
+        printf( "adding %s at %s at %d\n", e.type.c_str(), dir.c_str(), e.cur_level );
+    }
+    const std::string om_cur = overmap_buffer.ter( pos ).id().c_str();
+    expansions[ base_dir ] = parse_expansion( om_cur );
+    expansion_data &e = expansions[ base_dir ];
+    printf( "adding %s at %s at %d\n", e.type.c_str(), base_dir.c_str(), e.cur_level );
+    max_expansions = std::max( 1, expansions[ base_dir ].cur_level / 2 );
+}
+
+// upgrade levels
+bool basecamp::has_level( const std::string &type, int min_level, const std::string &dir )
```

Comment:
Query functions should generally be `const` as they don't mutate the object.

Reply:
indexing a map can be an assignment so some of these can't be const.

Reply:
Yes, but *why* are you indexing it (this also creates a default constructed value, is that intended as well)? Use `map::find` to get an iterator to an *existing* value (and handle the case when it does not exists).

Reply:
The thought was that indexing a map would be easier, but the find() idiom works well enough.  Fixed.